File: sorting-in-linear-time/golang/bucket_sort.go
package main

import (
	"fmt"
	"math/rand"
	"slices"
)

func sort(data []int) {
	table := make([][]int, slices.Max(data)/10+1)

	for _, d := range data {
		table[d/10] = append(table[d/10], d)
	}

	fmt.Println("--- table before sorting ---")
	for i, t := range table {
		fmt.Println(i, ":", t)
	}

	for i := range table {
		slices.Sort(table[i])
	}

	fmt.Println("--- table before after ---")
	for i, t := range table {
		fmt.Println(i, ":", t)
	}

	clear(data)
	data = data[:0]

	for i := range table {
		data = append(data, table[i]...)
	}
}

func main() {
	data := make([]int, 100)

	for i := range 100 {
		data[i] = rand.Intn(100) + 1
	}

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	sort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}

File: sorting-in-linear-time/python/bucket_sort.py
import random


def sort(data):
    table = [[] for _ in range(max(data) // 10 + 1)]

    for d in data:
        table[d // 10].append(d)

    print("--- table before sorting ---")
    for i, t in enumerate(table):
        print(i, ":", t)

    for t in table:
        t = t.sort()

    print("--- table after sorting ---")
    for i, t in enumerate(table):
        print(i, ":", t)

    data.clear()
    for t in table:
        data.extend(t)


if __name__ == "__main__":
    data = [random.randint(1, 100) for _ in range(100)]

    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)
